### PR TITLE
[SPARK-29235][ML][Pyspark]Support avgMetrics in read/write of CrossValidatorModel

### PR DESCRIPTION
**Pull request by shahidki31**
### What changes were proposed in this pull request?
 Currently pyspark doesn't write/read `avgMetrics` in `CrossValidatorModel`, whereas scala supports it.

### Why are the changes needed?
 Test step to reproduce it:

```
dataset = spark.createDataFrame([(Vectors.dense([0.0]), 0.0),
     (Vectors.dense([0.4]), 1.0),
     (Vectors.dense([0.5]), 0.0),
      (Vectors.dense([0.6]), 1.0),
      (Vectors.dense([1.0]), 1.0)] * 10,
     ["features", "label"])
lr = LogisticRegression()
grid = ParamGridBuilder().addGrid(lr.maxIter, [0, 1]).build()
evaluator = BinaryClassificationEvaluator()
cv = CrossValidator(estimator=lr, estimatorParamMaps=grid, evaluator=evaluator,parallelism=2)
cvModel = cv.fit(dataset)
cvModel.write().save("/tmp/model")
cvModel2 = CrossValidatorModel.read().load("/tmp/model")
print(cvModel.avgMetrics) # prints non empty result as expected
print(cvModel2.avgMetrics) # Bug: prints an empty result.
```

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Manually tested

Before patch:
```
>>> cvModel.write().save("/tmp/model_0")
>>> cvModel2 = CrossValidatorModel.read().load("/tmp/model_0")
>>> print(cvModel2.avgMetrics)
[]
```

After patch:
```
>>> cvModel2 = CrossValidatorModel.read().load("/tmp/model_2")
>>> print(cvModel2.avgMetrics[0])
0.5
```

### CODE REVIEW

**Comment by shahidki31** [comment]
cc @zhengruifeng Kindly review

**Review by srowen — APPROVED** [review]

**Comment by zhengruifeng** [comment]
```
In [1]: from pyspark.ml.classification import LogisticRegression

In [2]: from pyspark.ml.evaluation import BinaryClassificationEvaluator

In [3]: from pyspark.ml.linalg import Vectors

In [4]: dataset = spark.createDataFrame(
   ...:     ...     [(Vectors.dense([0.0]), 0.0),
   ...:     ...      (Vectors.dense([0.4]), 1.0),
   ...:     ...      (Vectors.dense([0.5]), 0.0),
   ...:     ...      (Vectors.dense([0.6]), 1.0),
   ...:     ...      (Vectors.dense([1.0]), 1.0)] * 10,
   ...:     ...     ["features", "label"]).repartition(1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-47bd70df4aa7> in <module>()
      1 dataset = spark.createDataFrame(
      2     ...     [(Vectors.dense([0.0]), 0.0),
----> 3     ...      (Vectors.dense([0.4]), 1.0),
      4     ...      (Vectors.dense([0.5]), 0.0),
      5     ...      (Vectors.dense([0.6]), 1.0),

TypeError: 'ellipsis' object is not callable

In [5]: dataset = spark.createDataFrame([(Vectors.dense([0.0]), 0.0),(Vectors.dense([0.4]), 1.0),(Vectors.dense([0.5]), 0.0),(Vectors.dense([0.6]), 1.0),(Vectors.dense([1.0]), 1.0)] * 10,["features", "label
   ...: "]).repartition(1)

In [6]: lr = LogisticRegression()

In [7]: grid = ParamGridBuilder().addGrid(lr.maxIter, [0, 1]).build()
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-7-045a988cd0ea> in <module>()
----> 1 grid = ParamGridBuilder().addGrid(lr.maxIter, [0, 1]).build()

NameError: name 'ParamGridBuilder' is not defined

In [8]: from pyspark.ml.tuning import *

In [9]: grid = ParamGridBuilder().addGrid(lr.maxIter, [0, 1]).build()

In [10]: evaluator = BinaryClassificationEvaluator()

In [11]: tvs = TrainValidationSplit(estimator=lr, estimatorParamMaps=grid, evaluator=evaluator, parallelism=1, seed=42)

In [12]: tvsModel = tvs.fit(dataset)
19/10/09 09:36:51 WARN BLAS: Failed to load implementation from: com.github.fommil.netlib.NativeSystemBLAS
19/10/09 09:36:51 WARN BLAS: Failed to load implementation from: com.github.fommil.netlib.NativeRefBLAS

In [13]: tvsModel.save("/tmp/model")

In [14]: tvsModel2 = TrainValidationSplitModel.load("/tmp/model")

In [15]: tvsModel.validationMetrics
Out[15]: [0.5, 0.8857142857142857]

In [16]: tvsModel2.validationMetrics
Out[16]: []
```

@shahidki31 Same issue also exist in `TrainValidationSplitModel`, can you also fix it in this pr?
BTW, what about adding doctests for model savle/load? (also check the loaded metrics)

**Comment by shahidki31** [comment]
Thanks @zhengruifeng  I will add metrics for `TrainValidationSplitModel` too.

**Comment by srowen** [comment]
If you'll make the changes @shahidki31 I think we can merge this.

**Comment by shahidki31** [comment]
Thanks @srowen . I will update it today. Actually, there seems an issue. I think `AvgMetrics` need to convert from java to python object, while reading.

**Comment by shahidki31** [comment]
Updated the PR. Locally verified.

**Comment by shahidki31** [comment]
retest this please

File: python/pyspark/ml/tuning.py [hunk 1]
```diff
@@ -829,13 +850,11 @@ def _to_java(self):
...
         """
 
-        sc = SparkContext._active_spark_context
-        # TODO: persst validation metrics as well
         _java_obj = JavaParams._new_java_obj(
             "org.apache.spark.ml.tuning.TrainValidationSplitModel",
             self.uid,
             self.bestModel._to_java(),
-            _py2java(sc, []))
+            self.validationMetrics)
```
Comment (srowen) [inline python/pyspark/ml/tuning.py]: This seems fine but out of curiosity why is the `_py2java` call no longer needed here?

Comment (shahidki31) [reply python/pyspark/ml/tuning.py]: I think we will be converting _py2java here?
https://github.com/apache/spark/blob/9e42c52c7728250547aa044f8c28f93de9940460/python/pyspark/ml/wrapper.py#L60-L69

Comment (shahidki31) [reply python/pyspark/ml/tuning.py]: I compared with _py2java here and without here, both cases the written metadata  file is same. I'll add _py2java here, for consistency.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by shahidki31** [comment]
Thanks @srowen @zhengruifeng

**Pull request closed without merging**